File: src/creational-patterns/factory/factory.rb
# frozen_string_literal: true

# An implementation of the Factory design pattern
class ProductFactory
  private_class_method :new

  def initialize
    @registered_products = {}
  end

  def self.instance
    @instance ||= new
    @instance
  end

  def register_product(product_id, product_class)
    @registered_products[product_id] = product_class
  end

  def create_product(product_id)
    if @registered_products.include?(product_id)
      @registered_products[product_id].new('test')
    else
      throw ArgumentError 'Invalid product ID'
    end
  end
end

# The objects that will be created by the Factory class
class Product
  def initialize(name)
    @name = name
  end

  def describe
    "This product is: #{@name}"
  end
end

# A more specific class that we will register with the factory
class SpecificProduct < Product
  ProductFactory.instance.register_product('SPECIFIC', SpecificProduct)

  def initialize(name)
    name = "Specific #{name}"
    super(name)
  end
end

File: src/creational-patterns/factory/factory.js
const singleton = Symbol('singleton');
const singletonEnforcer = Symbol('enforcer');

class ProductFactory {
  constructor(enforcer) {
    if (enforcer !== singletonEnforcer) return null;
    this.registeredProducts = {};
  }

  registerProduct(productId, productClass) {
    this.registeredProducts[productId] = productClass;
  }

  createProduct(productId) {
    let product = null;
    if (this.registeredProducts[productId]) {
      product = new this.registeredProducts[productId]('test');
    }
    return product;
  }

  static get instance() {
    if (!this[singleton]) {
      this[singleton] = new ProductFactory(singletonEnforcer);
    }
    return this[singleton];
  }
}

module.exports = ProductFactory;
